README.md
The 'OptCirClust' R package
=================================

### Overview

The package provides fast, optimal, and reproducible clustering for circular, periodic, or framed data. The performance is guaranteed by embedding divide-and-conquer, dynamic programming, and another divide-and-conquer. The runtime of the algorithms is $O(K N \log^2 N)$, where $K$ is the number of clusters and $N$ is the number of circular data points.

The core algorithm solves the framed clustering problem by minimizing the within-cluster sum of squared distances. In contrast to heuristic clustering, the efficiency and accuracy of fast optimal framed clustering is evident both theoretically and practically. The fast algorithm use divide-and-conquer inside dynamic programming inside another divide-and-conquer, guaranteeing cluster optimality---the total of within-cluster sums of squared distances is always the minimum given the number of clusters. Two plot functions visualize circular and framed clusters to reveal patterns in data. On a desktop computer using a single processor core, millions of data points can be clustered within seconds. The algorithms offer a general high-performance solution to circular, periodic, or framed data clustering.

### Main methods

The function to perform fast optimal circular data clustering is `CirClust()`. The function to perform the fast framed clustering is `FramedClust()`. Both functions with default `method` have a runtime of $O(K N \log^2N)$, linear polylogarithmic in sample size $N$ and linear in the number of clusters $K$. The space complexity of both functions is $O(KN)$. Their implementations scale large input data to improve numerical precision.

### When to use the package

One can use the algorithms to characterize events along circular DNA molecules, circular RNA molecules, and circular genomes of bacteria, chloroplast, and mitochondria. One can also cluster climate data along any given longitude or latitude. Periodic data clustering can be formulated as circular clustering.

### To download and install the package
```{r}
install.packages("OptCirClust")
```


NEWS.md
# NEWS


## Version 0.0.2

    2020-12-12
  
  1. Fixed an issue of missing the C++ header file <limits> in the OptFramedClust.h.
  2. Removed unused C++ header files.
  3. Add three more parameters (fill, border, border.lty) to give the user more control over visualizing circular data clustering results by function plot.CirClust().
  4. Updated vignettes and manuals.
  
  
    2020-12-09
  
  1. Version 0.0.2 was created from version 0.0.1
  2. Return value of CirClust are now fully documented in the manual according to the code.
  3. Checked the return value of FramedClust. 
  

## Version 0.0.1
    
    2020-12-07
    
  1. Updated plot.CirClust() function to avoid calling `par()` function. 
     This will give the user full control of graphic parameters.
     Updated the function's manual, examples, and related vignettes. 
  2. Updated package DESCRIPTION and README.md.
  3. Changed package title.
  
    
    2020-12-04
    
  1. Included with the package a FASTA file downloaded from
     GenBank accession number CP019943.1, representing the
     circular genome of Candidatus Carsonella ruddii, used
     in the vignette "Circular genome clustering".
  
  
    2020-12-03
    
  1. Fixed issues with example(FramedClust).
  2. Optimized plot.CirClust output  
    
    
    2020-12-01
  1. Fixed an issue where FramedClust crashed if the first.frame is
     not 1. E.g., 
        FramedClust(1:10, 2, 4, 2, 7)
        FramedClust(1:10, 2, 4, 2, 6)

    
    2020-11-30
    
  1. The orange arrows in the "Circular genome clustering" vignette
     are visible now.
  2. Changed the frame.width argument in FramedClust function to
     frame.size, to reduce potential confusion.
  
    
    2020-11-29
    
  1. In FramedClust, replaced the default 
        last.frame = length(X)-frame.width
     by 
        last.frame = length(X)-frame.width+1,
  
  2. Replaced the circular data examples in plot.FramedClust
  3. Defined the meaning of frame.width in the user document and the
  details section.
  4. Updated the FramedClust() manual. 
     + explained the NA in the 'cluster' component
     + removed the 'ID' component. 
     + removed "best.frame -> ID", and "Clearly define Border".
      
        
    2020-11-24

  1. Updated the package title and description. 
  
  
    2020-10-06
  
   1. Created a unified function for framed clustering 
  FramedClust(
    X, K, frame.width,     
    first.frame = 1,
    last.frame = length(X)-frame.width+1, 
    method = c("linear.polylog", "kmeans", "Ckmeans.1d.dp")
  )
  
<!--  2. create an R interface to lin_polylog_framed_clust 
  
  (do we need this seperate function, we are handeling all the framed clusters in the FramedClust function)
  
  lin.polylog.framed.clust(X, K, frame.width, first.frame, last.frame), inside which you call lin_polylog_framed_clust()
  
  For lin.polylog.framed.clust, the indices must be 1-based to be consistent with R;
  For lin_polylog_framed_clust, the indices must be 0-based to be consistent with C/C++.
-->

<!--  
    2020-09-29
  
  1. Do not export lin_polylog_framed_clust(), 
     kmeans.framed.clust(), quad.framed.clust() functions. Thus, we do not have to maintain three manuals that look almost identical.
     
  2. Reorder/rename the arguments to
  
  + lin_polylog_framed_clust(X, K, frame_width, first_frame, last_frame, prev, next)
  + kmeans.framed.clust(X, K, frame.width, first.frame, last.frame)
  + quad.framed.clust(X, K, frame.width, first.frame, last.frame) 
-->


    2020-09-13
  
  1. Completed the manual for function lin_polylog_framed_clust.
  2. Wrote the description of the package.
  3. Created a README.md file.
  4. Tested the kmeans.framed.clust() function.
    
    
    2020-08-11
    
  1. Package name changed to OptCirClust to avoid confusion with a
     similar R package.
  
    2020-08-10
  
  1. Added more tests for function CirClust(), such as irregular input data.
  2. Changed the for-loops in CirClust() to vectorized operations.
  3. Corrected the mean in OCC.
  
  
    2020-08-08
  
  1. Created CirClust version 0.0.1 from optCclust version 0.1.0
  2. Renamed function Cir.Clust() to CirClust() to reduce confusion.
     This function returns an object of class "CirClust".
  3. Renamed C/C++ function FOCC_Interface() to 
     lin_polylog_framed_clust().
  4. Renamed file "FOCC_Interface.cpp" to 
     "lin_polylog_framed_clust.cpp"
  5. Exported lin_polylog_framed_clust as a function for the package.

<!--  
    2020-08-08
  
  1. Shall we remove the following two arguments  (Yes)
       prev_k_f = -1,
       next_k_f = -1
       from FramedClust? I would think the user will never
       use them. Is so, will they cause confusion to the user?
-->  
  


R/CirClust.R
# CirClust.R
#'
#'
#'
#' Circular Data Clustering
#'
#' Perform clustering on circular data to minimize the
#'   within-cluster sum of squared distances.
#'
#' @useDynLib OptCirClust, .registration = TRUE
#' @importFrom Rcpp sourceCpp
#'
#' @param O a vector of circular data points. They can be coordinates
#'   along the circle based on distance, or angles around the circle.
#'
#' @param K the number of clusters
#'
#' @param Circumference the circumference of the circle where
#'   data are located
#'
#' @param method the circular clustering method.
#' \code{"FOCC"}: fast and optimal, the default method;
#'   \code{"HEUC"}: based on heuristic k-means, fast but not necessarily optimal;
#'   \code{"BOCC"}: brute-force based on Ckmeans.1d.dp, slow but optimal,
#'   included to provide a baseline.
#'
#'
#' @details In clustering \eqn{N} circular points into \eqn{K} clusters, the "FOCC" algorithm
#' is reproducible with runtime \eqn{O(K N \log^2 N)}{O(K N log^2 N)};
#' The "HEUC" algorithm, not always reproducible, calls the \code{kmeans} function repeatedly;
#' The "BOCC" algorithm with runtime \eqn{O(KN^2)}, reproducible but slow, is done via
#' repeatedly calling the \code{Ckmeans.1d.dp} function.
#'
#'
#'
#' @return An object of class \code{"CirClust"} which has a \code{plot}
#'  method. It is a list with the following components:
#'
#'
#'
#' \item{ID}{the starting index of the frame with minimum SSQ}
#'
#' \item{Border}{the borders of \code{K} clusters}
#'
#' \item{cluster}{a vector of clusters assigned to each element in \code{O}.
#' Each cluster is indexed by an integer from 1 to \code{K}.}
#'
#' \item{centers}{a numeric vector of the means for each cluster in the circular data.}
#'
#' \item{withinss}{a numeric vector of the within-cluster sum of squares for each cluster.}
#'
#' \item{size}{a vector of the number of elements in each cluster.}
#'
#' \item{totss}{the total sum of squared distances between each element and the sample mean.
#' This statistic is not dependent on the clustering result.}
#'
#' \item{tot.withinss}{the total sum of within-cluster squared distances between
#'  each element and its cluster mean. This statistic is minimized given the number of clusters.}
#'
#' \item{betweenss}{the	sum of squared distances between each cluster mean and sample mean.
#' This statistic is maximized given the number of clusters.}
#'
#' \item{O_name}{a character string. The actual name of the \code{O} argument.}
#'
#'
#' @examples
#' O <- c(1,2, 10,11,12,13,14,15, 27,28,29,30,31,32, 40,41)
#'
#' K <- 3
#'
#' Circumference <- 42
#'
#' # Perform circular clustering:
#' output <- CirClust(O, K, Circumference)
#'
#' # Visualize the circular clusters:
#' opar <- par(mar=c(1,1,2,1))
#' plot(output)
#' par(opar)
#'
#'
#' @export
#'
CirClust <- function(O,
                     K,
                     Circumference,
                     method = c("FOCC", "HEUC", "BOCC"))
{
  O_name <- deparse(substitute(O))

  method <- match.arg(method)

  O <- O %% Circumference

  if (FALSE)
  {
    if (!identical(which(O < 0), integer(0)))
    {
      O[which(O < 0)] <- O[which(O < 0)] + Circumference
    }

    if (!identical(which(O > Circumference), integer(0)))
    {
      O[which(O > Circumference)] <-
        O[which(O > Circumference)] - Circumference
    }
  }




  frame.width <- length(O)

  I <- order(O)

  # O_sort <- sort(O, decreasing = FALSE)

  O_sort <- O[I]

  X <- c(O_sort, (O_sort + Circumference))

  first.frame <- 0

  prev_k_f <- -1

  next_k_f <- -1

  last.frame <- length(X) - frame.width - 1

  if (method == "FOCC")
  {
    result <-
      lin_polylog_framed_clust(
        as.double(X), as.integer(K), as.integer(frame.width),
        as.integer(first.frame), as.integer(last.frame),
        as.integer(prev_k_f), as.integer(next_k_f))

    cluster <-  rep(1, (result$Border[1] - result$ID + 1))

    if(K > 1)
    {
      for (i in 2:K)
      {
        cluster <-
          c(cluster, rep(i, (result$Border[i] - result$Border[i - 1])))
      }
    }


  }
  else if (method == "BOCC")
  {
    result <- quad.framed.clust(X, K, frame.width, first.frame, last.frame )

    cluster <- result$cluster

    size <- result$size[unique(result$cluster)]

    result$Border <- cumsum(size) + result$ID - 1

  }
  else if (method == "HEUC")
  {
    result <- kmeans.framed.clust(X, K, frame.width, first.frame, last.frame)

    cluster <- result$cluster

    size <- result$size[unique(result$cluster)]

    result$Border <- cumsum(size) + result$ID - 1
  }










  if (result$ID > 0)
  {
    cluster_new <-
      c(cluster[(frame.width - result$ID + 1):frame.width], cluster[1:(frame.width - result$ID)])    # The result$ID count starts from zero

  }
  else
  {
    cluster_new <- cluster
  }




  cluster[I] <- cluster_new


  if (FALSE)
  {
    count = 1


    while (count <= length(O_sort))
    {
      cor <- which(O %in% O_sort[count])



      cluster[cor] <- cluster_new[count]
      count = count + length(cor)


    }
  }


  result$centers <- result$centers %% Circumference


  if (FALSE)
  {
    if (!identical(which(result$centers > Circumference), integer(0)))
    {
      result$centers[which(result$centers > Circumference)] <-
        result$centers[which(result$centers > Circumference)] - Circumference
    }

    if (!identical(which(result$centers < 0), integer(0)))
    {
      result$centers[which(result$centers < 0)] <-
        result$centers[which(result$centers < 0)] + Circumference
    }
  }



  # result$centers <- sort(result$centers, decreasing = FALSE)


  Border.mid <- (X[result$Border + 1] + X[result$Border + 2]) / 2

  result_new <-
    list(
      "cluster" = cluster,
      "centers" = result$centers,
      "withinss" = result$withinss,
      "size" = result$size,
      "totss" = result$totss,
      "tot.withinss"  = result$tot.withinss,
      "betweenss" = result$betweenss,
      "ID" = result$ID,
      "Border" = result$Border,
      "Border.mid" = Border.mid,
      "O_name" = O_name,
      "Circumference" = Circumference
    )

  class(result_new) <- "CirClust"

  return(result_new)
}
